src/NumberInput.jsx

```diff
@@ -56,6 +63,23 @@ export default React.createClass({
       this.getDefaultState(nextProps))
   },
 
+  componentWillUpdate(nextProps, nextState) {
+    if (canUseDOM) {
+      // Check if we need to restore text selection after pending update.
+      let node = ReactDOM.findDOMNode(this.refs.input);
+      this._selectAll = document.activeElement === node
+        && this.state.stringValue !== nextState.stringValue
```

Comment:
Sorry for the delayed reply.

I did some more testing, and I think it should be safe, with one exception. For the case of pasting, for example, the paste updates the field's contents and changes selectionStart, selectionEnd to the end of the text, then the onChange event fires, then componentWillUpdate and componentDidUpdate fire. So, by the time componentWillUpdate and compomentDidUpdate are fired, the selection is already at the end of the text, so the _selectAll behavior isn't triggered.

The only exception I've been able to find is if you trigger the spin button behavior by pressing Up or Down. In that case, the prop or state is changed directly, so all text is still selected when componentDidUpdate is fired, and the _selectAll behavior is triggered. Having the spin buttons preserve selection isn't necessarily bad, but it still feels a little odd, and it's inconsistent with the behavior of the spin buttons when some but not all text is selected.

The easiest solution, I think, would be to have NumberPicker.spin clear any selection before calling NumberPicker.change. I can add this logic if you're still open to this PR.

I posted some demo code I was using to test all of this to https://github.com/joshkel/react-widgets-tab.